File: 14-TypeScript/Grey Multi-Tenant/packages/grey-haskell/src/Grey.hs
-- |
-- Module      : Grey
-- Description : Grey Multi-Tenant SDK for Haskell
-- License     : MIT
-- Stability   : experimental
--
-- The Grey SDK provides a pure functional interface to the Grey Multi-Tenant
-- API using gRPC transport.
--
-- = Quick Start
--
-- @
-- import Grey
--
-- main :: IO ()
-- main = do
--   -- Create a local client
--   client <- localClient Nothing
--
--   -- Login
--   result <- Auth.login (grpcClient client) credentials
--   case result of
--     Ok tokens -> putStrLn $ "Logged in: " ++ show (accessToken tokens)
--     Err e -> putStrLn $ "Error: " ++ show e
--
--   -- Cleanup
--   closeClient client
-- @
--
-- = Error Handling
--
-- All operations return a 'Result' type which is either 'Ok' with a value
-- or 'Err' with a 'GreyError'.
--
-- @
-- result <- Users.getUser grpc "user-123"
-- case result of
--   Ok user -> print user
--   Err e -> do
--     putStrLn $ "Error code: " ++ show (errorCode e)
--     putStrLn $ "Message: " ++ show (errorMessage e)
--     when (isRetryable e) $ putStrLn "This error is retryable"
-- @

module Grey
  ( -- * Client
    GreyClient(..)
  , newClient
  , localClient
  , productionClient
  , closeClient
  , withClient
  , withAuthToken
  , withTimeout
  , withHeader

    -- * Configuration
  , Options(..)
  , defaultOptions
  , localOptions
  , productionOptions

    -- * Error Handling
  , Result(..)
  , ok
  , err
  , isOk
  , isErr
  , fromResult
  , getOr
  , mapResult
  , mapError
  , flatMap

    -- * Error Types
  , GreyError(..)
  , ErrorCode(..)
  , isRetryable

    -- * Domain Clients
  , module Grey.Domain.AuthClient
  , module Grey.Domain.UserClient
  , module Grey.Domain.ProjectsClient
  , module Grey.Domain.QueryClient
  , module Grey.Domain.MutationClient

    -- * gRPC Client
  , GrpcClient
  , grpcClient
  ) where

import Grey.Client
import Grey.Config.Options
import Grey.Error.Codes (ErrorCode(..))
import Grey.Error.GreyError (GreyError(..), isRetryable)
import Grey.Error.Result
import Grey.GRPC.Client (GrpcClient)

import Grey.Domain.AuthClient
import Grey.Domain.UserClient
import Grey.Domain.ProjectsClient
import Grey.Domain.QueryClient
import Grey.Domain.MutationClient

File: 14-TypeScript/Grey Multi-Tenant/packages/grey-zig/src/grey.zig
//! Grey Multi-Tenant SDK for Zig
//!
//! A complete SDK for integrating with Grey Multi-Tenant services using
//! gRPC transport via C interop.
//!
//! ## Quick Start
//!
//! ```zig
//! const grey = @import("grey");
//!
//! pub fn main() !void {
//!     var gpa = std.heap.GeneralPurposeAllocator(.{}){};
//!     defer _ = gpa.deinit();
//!     const allocator = gpa.allocator();
//!
//!     // Create client
//!     var client = try grey.createClient(allocator);
//!     defer client.deinit();
//!
//!     // Login
//!     const login_result = client.auth.login("user@example.com", "password123");
//!     if (login_result.isOk()) {
//!         std.debug.print("Logged in successfully!\n", .{});
//!     }
//! }
//! ```

const std = @import("std");

// Error types
pub const error_codes = @import("error/codes.zig");
pub const grey_error = @import("error/grey_error.zig");
pub const result = @import("error/result.zig");

pub const ErrorCode = error_codes.ErrorCode;
pub const GreyError = grey_error.GreyError;
pub const Result = result.Result;
pub const VoidResult = result.VoidResult;

// Configuration
pub const options = @import("config/options.zig");
pub const Options = options.Options;

// gRPC layer
pub const grpc_client = @import("grpc/client.zig");
pub const auth_service = @import("grpc/auth_service.zig");
pub const user_service = @import("grpc/user_service.zig");
pub const projects_service = @import("grpc/projects_service.zig");
pub const query_service = @import("grpc/query_service.zig");
pub const mutation_service = @import("grpc/mutation_service.zig");

pub const Channel = grpc_client.Channel;

// Domain clients
pub const auth_client = @import("domain/auth_client.zig");
pub const user_client = @import("domain/user_client.zig");
pub const projects_client = @import("domain/projects_client.zig");
pub const query_client = @import("domain/query_client.zig");
pub const mutation_client = @import("domain/mutation_client.zig");

pub const AuthClient = auth_client.AuthClient;
pub const UserClient = user_client.UserClient;
pub const ProjectsClient = projects_client.ProjectsClient;
pub const QueryClient = query_client.QueryClient;
pub const MutationClient = mutation_client.MutationClient;

// Main client
pub const client = @import("client.zig");
pub const GreyClient = client.GreyClient;

/// Create a new Grey client with default local options
pub const createClient = client.createClient;

/// Create a new Grey client with custom options
pub const createClientWithOptions = client.createClientWithOptions;

// Version info
pub const version = "0.0.1";
pub const version_major: u32 = 0;
pub const version_minor: u32 = 0;
pub const version_patch: u32 = 1;

test {
    // Run all tests
    std.testing.refAllDecls(@This());
}
